activerecord/lib/active_record/query_constraints.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module QueryConstraints
+    extend ActiveSupport::Concern
+
+    included do
+      class_attribute :query_constraints_list, instance_writer: false
+    end
+
+    module ClassMethods
+      # Accepts a list of attribute names to be used in the WHERE clause
+      # of SELECT / UPDATE / DELETE queries.
+      #
+      #   class Developer < ActiveRecord::Base
+      #     query_constraints :company_id, :id
+      #   end
+      #
+      #   developer = Developer.first
+      #   developer.inspect # => #<Developer id: 1, company_id: 1, ...>
+      #
+      #   developer.update!(name: "Nikita")
+      #   # => UPDATE "developers" SET "name" = 'Nikita' WHERE "developers"."company_id" = 1 AND "developers"."id" = 1
+      #
+      #   It is possible to update attribute used in the query_by clause:
+      #   developer.update!(company_id: 2)
+      #   # => UPDATE "developers" SET "company_id" = 2 WHERE "developers"."company_id" = 1 AND "developers"."id" = 1
+      #
+      #   developer.name = "Bob"
+      #   developer.save!
+      #   # => UPDATE "developers" SET "name" = 'Bob' WHERE "developers"."company_id" = 1 AND "developers"."id" = 1
+      #
+      #   developer.destroy!
+      #   # => DELETE FROM "developers" WHERE "developers"."company_id" = 1 AND "developers"."id" = 1
+      #
+      #   developer.delete
+      #   # => DELETE FROM "developers" WHERE "developers"."company_id" = 1 AND "developers"."id" = 1
+      #
+      #   developer.reload
```

Comment:
Just a note that `reload` doesn't work as intended at the moment as it wasn't relying on `_primary_key_constraints_hash` and a follow-up PR incoming to fix this